pipeline_dp/dp_computations.py

```diff
@@ -247,3 +254,56 @@ def compute_dp_mean(count: int, sum: float, dp_params: MeanVarParams):
                                  dp_params.noise_kind)
     dp_mean = dp_normalized_sum / dp_count + middle
     return dp_count, dp_mean * dp_count, dp_mean
+
+
+def compute_dp_var(count: int, sum: float, sum_squares: float,
```

Comment:
There's another trick we can use for even more efficient variance calculations. The variance of [x_0, x_1, ..., x_n] is the same as the variance of [x_0 - k, x_1 - k, ... , x_n - k]. In other words, the variance doesn't change when you translate all the elements in the dataset by a fixed amount.

Right now you're calculating the mean by normalizing all the elements to the midpoint, and calculating the mean of squares by normalizing all elements to the midpoint of squares. E.g.

for each input:
  sum += input - midpoint
  sum_of_squares += input ** 2 - midpoint_of_squares(lower, upper)

noisy_sum = add_noise(sum)
noisy_sum_of_squares = add_noise(sum)

noisy_mean = noisy_sum / noisy_count + midpoint
noisy_mean_of_squares = noisy_sum_of_squares / noisy_count + midpoint_of_squares(lower, upper)

Instead, we can translate the elements so they're centered around 0 before doing the rest of the calculation, e.g.

for each input:
  centered_input = input - midpoint
  sum += centered_input  # The midpoint is already 0.
  sum_of_squares += centered_input ** 2 - midpoint_of_squares(lower - midpoint, upper - midpoint)

  noisy_sum = add_noise(sum)
  noisy_sum_of_squares = add_noise(sum_of_squares)
  
  noisy_mean = noisy_sum / noisy_count
  noisy_mean_of_squares = noisy_sum_of_squares / noise_count + midpoint_of_squares(lower - midpoint, upper - midpoint)

Then we use those values to calculate the variance (note we don't un-normalize the mean). We can also post-process the noisy mean to shift it back to the original range. We probably don't need to implement that in this version, but it's something to consider adding in the future.

Reply:
Thanks for the review! :) About these more efficient calculations, should I create a new issue/pull request with the implementation?

Reply:
Yeah, let's do it in the subsequent CL